lib/active_model/serializer/adapter/json_api/deserialization.rb

```diff
@@ -0,0 +1,120 @@
+module ActiveModel
+  class Serializer
+    module Adapter
+      class JsonApi
+        module Deserialization
+          module_function
+
+          # Parse a Hash or ActionController::Parameters representing a JSON API document
+          # into an ActiveRecord-ready hash.
+          #
+          # @param [Hash|Object implementing `to_h`] document
+          # @param [Hash] options
+          #   fields: Array of symbols and a Hash. Specify whitelisted fields, optionally
+          #     specifying the attribute name on the model.
+          # @return [Hash] ActiveRecord-ready hash
+          #
+          def parse(document, options = {})
+            fields = parse_fields(options[:fields])
+
+            hash = {}
+
+            document = document.to_h unless document.is_a?(Hash)
+            primary_data = document.fetch('data', {})
+            included_data = document.fetch('included', {})
+
+            hash.merge!(parse_primary_data(primary_data, fields))
+            hash.merge!(parse_included_data(included_data, fields))
+
+            hash
+          end
+
+          def parse_included_data(included_data, fields)
+            data = {}
+            included_data.each do |included_resource|
+              parsed = parse_primary_data(included_resource, fields)
+              # parsed should only have one key
+              # so there is no possibility of missing other keys when
+              # calling first
+              # NOTE: key is always going to be singular
+              key = parsed.keys.first
+              plural_key = key.pluralize
+              parsed_attributes = parsed[key]
+              # on multiple iterations of parsing the same type
+              # the key in data is changed to be plural,
+              # so we need to check for the existence of both
+              # singular and plural keys
+              exists = data.keys.include?(key) || data.keys.include?(plural_key)
```

Comment:
@beauby I'm pro @NullVoxPopuli and @vasilakisfil If the code isn't self-documenting, it needs documentation.  Comments help tremendously.  Ideal could is self documenting, real code, whether good or bad, may benefit from comments explaining why the code is the way it is, what decisions were made, and why.  When I read code, I don't want to have to pull my hair out trying to figure out what it does. So, major props to @NullVoxPopuli for pointing out this example